### PR TITLE
Preserve unit type in `toBest`

### LINKED ISSUE #318: Result of `toBest` should preserve unit type
We use `convert-units` and instantiate it something like this:

```typescript
import configMeasurements, { BestResult, Measure } from "convert-units";

// stock units
import mass, { MassUnits } from "convert-units/definitions/mass";
import volume, { VolumeUnits } from "convert-units/definitions/volume";

// our custom units
import density, { DensityUnits } from "./density";

type OurUnits = DensityUnits | MassUnits | VolumeUnits;

export const convert = configMeasurements<
  "density" | "mass" | "volume",
  "imperial" | "metric",
  OurUnits // <-- nicely typed!
>({
  density,
  mass,
  volume,
});
```

It would be nice if, when we call `convert(value).from(fromUnits).toBest(...)` and the result is non-null, the `unit` field on the result could be known to have type `OurUnits` instead of just `string`. That should always be the case, but currently we need a manual `as`-cast.

I think this is easy and can be done in a backward-compatible manner, so I'll send a PR shortly.

### PR DESCRIPTION
**Pull request by wchargin**
The `BestResult` type now takes a type parameter `TUnits`, which defaults to `string` for backward compatibility. That way, users who have configured a measurement system with a custom unit type can preserve that unit type on `bestResult.unit`.

The code was already nicely set up for this, so no casts or refactors were needed, just enriching the types themselves.

Fixes #318.

Test Plan:
Type-level unit test added; running `npm test` passes with this change but fails if you revert the change to `src/convert.ts` only. It also fails if you change `actual?.unit` to `'wat'`, with an appropriate error message, so it's testing the right thing and is not vacuous.

wchargin-branch: tobest-preserve-unit-type

### CODE REVIEW

File: src/__tests__/best.test.ts
Comment (Taar) [inline src/__tests__/best.test.ts:17]: This should be in it's own test rather than placed in an existing test. If the test fails, it's less clear to why the test is failing given the test's description of `best mm`. Maybe a test with the description of `type of unit equals configured units` or something like that would be more descriptive.

While relying on typescript to fail the test is one way this could be tested, the test's code is rather confusion at first glance. I think the test would more obvious if a type guard was used to validate the type. Typescript would still catch any type errors as well as the test would fail if the string is anything but a `LengthUnits` in this case. This would also remove the need to suppress the `eslint/no-unused-vars` rule.  

Let me know your thoughts on this and thanks again for submitting a PR.

**Review by Taar — CHANGES_REQUESTED** [review]
Thanks for the PR! I agree that this is a nice improvement to the `toBest` method. My only concern is the test which I've a left comment on. Once that is addressed I'd be happy to merge this PR.

File: src/__tests__/best.test.ts
Comment (wchargin) [reply src/__tests__/best.test.ts]: Thanks for the quick review! I agree about moving it to its own test: done.

I don't quite understand what you're suggesting about type guards. Are you talking about [type guards] in the sense of "functions whose return type is something like `arg is SomeType`"? We could write a type guard in the test case like this—

```typescript
function isLengthUnit(value: string): value is LengthUnits {
  const lengthUnits = Object.values(length.systems).flatMap((system) =>
    Object.keys(system)
  );
  return lengthUnits.includes(value);
}

test('type of unit field matches configured units', () => {
  const convert = configureMeasurements<'length', LengthSystems, LengthUnits>({
    length,
  });
  const best = convert(1200).from('mm').toBest();
  expect(isLengthUnit(best!.unit)).toBe(true);
});
```

—but this doesn't actually test what's desired. The test will pass (both at runtime and `tsc`) even if you revert all the changes to `convert.ts` in this patch, because type guards are used to ask TypeScript to assume that a typing judgment holds, rather than asking TypeScript to validate a typing judgment.

I switched to a version that is a bit more explicit and also does not contain any lint suppressions. It still relies on TypeScript for the checking, but I think that's okay; fundamentally, this is a test about TypeScript behavior, not runtime behavior. If I'm misunderstanding your suggestion, please let me know and I'd be happy to change it to better fit your preferences.

[type guards]: https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

Comment (Taar) [reply src/__tests__/best.test.ts]: Oof, Sorry for all the confusion! I should have waited until I had more time to think about my response and test a few things out. I thought maybe that a type guard could be used to test both the type and the runtime value which, imo, would make the test's purpose a lot more obvious. I see now that using a type guard would not help achieve that goal.

I agree with you that it's fine to rely on Typescript to test the type. I don't think it's possible to test a type in any other context. However, I still don't like how the test is defined and after a bit more thought, I think it's the mixing of type and runtime checks that are making me pause. Sorry if that statement is a bit confusing. I'm struggling to find the right words to explain my thoughts so please let me know if things are still unclear.

First off, I don't think the test is actually testing the right type. It would pass if the type is `LenghtUnits | undefined` while the `toBest` has the return type of `BestResult<TUnits> | null`. I think removing the whole runtime statement should partly solve this problem and extracting the return type using `ReturnType`. This would remove the code that I think would be a major source of confusing for anyone trying to understand the test.  In fact, I think that there shouldn't be any runtime code used in the test at all. The following code hopefully shows how this would be possible:

```typescript
test('toBest return type should equal `BestResult<LengthUnits> | null`', () => {
  type toBestMethod = Converter<'length', LengthSystems, LengthUnits>['toBest'];
  type returnTypeOfToBest = ReturnType<toBestMethod>;
});
```

I think the above code snippet makes it very clear that the test is only dealing with types. The next step would be actually getting Typescript to error if the type is not what is expected.

While the `assertStaticType` function does get Typescript to throw an error. I'm still not sold on its implementation as it still feels a bit implicit and odd to me. I think it would be better to use something like [expect-type](https://github.com/mmkal/expect-type) to help test the type as it provides a lot more features and the API is a bit more descriptive/explicit to me. With all that said, I think the test could look something like this:

```typescript
import { expectTypeOf } from 'expect-type';
test('toBest return type should equal `BestResult<LengthUnits> | null`', () => {
  type toBestMethod = Converter<'length', LengthSystems, LengthUnits>['toBest'];
  expectTypeOf<toBestMethod>().returns.toEqualTypeOf<BestResult<LengthUnits> | null>();
});
```

I hope things are bit more clear and I would love to know your thoughts on this! I would also like to encourage you to spend some time playing around with this to see if you can come up with a better solution. I'm also not completely sold on [expect-type](https://github.com/mmkal/expect-type). I chose it because its API is similar to jest's API. Feel free to propose alternatives :)

Comment (wchargin) [reply src/__tests__/best.test.ts]: Thanks for your response; this makes sense and I understand, and I agree with a lot of it.

> First off, I don't think the test is actually testing the right type. It would pass if the type is `LenghtUnits | undefined` while the `toBest` has the return type of `BestResult<TUnits> | null`.

This seems fair; I am happy to make the test more specific in this way.

> I hope things are bit more clear and I would love to know your thoughts on this!

It's clear, yes, and thank you. Since you asked for my thoughts, I will share two.

First, about reducing runtime code in the test: I agree with this in moderation. I do think that having *some* runtime code can make it easier for the test to match what an actual user would write. Conceptually, I'm trying to test, "when I call `toBest()` on a unit system with `LengthUnits`, the `unit` field on the result (when present) has type `LengthUnits`". Starting off that test by actually calling `toBest()` and checking the `unit` field feels simple and easy to read—simpler than messing with TypeScript utility types to transform and destructure the library types themselves, and far easier to tell that it's actually doing the same thing as a real user. But I think there's reasonable room to land in different places here, and if you place a high value on reducing runtime code, I can work with that.

Second, on the overall question of using something like `expectTypeOf`: I think really this is far more complicated and hard to understand than the alternative. In the version of your comment, we have a new external dependency whose API and mechanisms I need to read and understand. I need to ask questions like, is the generic type parameter on `expectTypeOf<_>` being inferred as strictly as it should be (but no stricter)? How does this interact with Jest's `.returns` operator at a typing level? And it *does* actually execute code at runtime; what does `expectTypeOf().toEqualTypeOf()` do? Does it try any run-time type-checking, too, or do these functions have empty bodies?

But most importantly, as part of this second point, is `BestResult<LengthUnits>` even the right type to check? It looks good on the surface, but note that if `BestResult<T>` still defines `unit: string`, **the test in your comment erroneously passes!** I think by trying to remove as much runtime code as possible, the test has simultaneously become more complicated and also less powerful, because it moves away from testing in the user's shoes and toward repeating the package's implementation.

I tried to fix this problem and write a correct version of the test with `expectTypeOf`. Here was my first draft:

```typescript
import { expectTypeOf } from 'expect-type';
import configureMeasurements, { type Converter } from '..'; // more type imports now, too

test('toBest return type should equal `BestResult<LengthUnits> | null`', () => {
  type toBestMethod = Converter<'length', LengthSystems, LengthUnits>['toBest'];
  type toBestResultWhenPresent = Exclude<ReturnType<toBestMethod>, null>;
  expectTypeOf<toBestResultWhenPresent>().toEqualTypeOf<LengthUnits>(); // oops
});
```

Oops! I wrote `expectTypeOf<toBestResultWhenPresent>` rather than `expectTypeOf<toBestResultWhenPresent['unit']>`. (Genuine mistake on my part; not contrived. It gets harder to read and write when you have so many high-power type manipulation utilities being thrown around!) What does this test failure look like? I get this error:

```
    src/__tests__/best.test.ts:283:57 - error TS2344: Type 'string' does not satisfy the constraint 
'{ [x: number]: "Expected: string, Actual: never"; length: "Expected: number, Actual: never"; [itera
tor]: "Expected: function, Actual: never"; val: "Expected: never, Actual: number"; ... 49 more ...; 
at: "Expected: function, Actual: never"; }'.
      Type 'string' is not assignable to type '{ [x: number]: "Expected: string, Actual: never"; len
gth: "Expected: number, Actual: never"; [iterator]: "Expected: function, Actual: never"; val: "Expec
ted: never, Actual: number"; ... 49 more ...; at: "Expected: function, Actual: never"; }'.

    283   expectTypeOf<toBestResultWhenPresent>().toEqualTypeOf<LengthUnits>(); // oops
```

Holy moly! "... 49 more ..."?! What even is going on here? I think we can both agree that we would rather debug a normal TypeScript error than that. TypeScript would report the same error as follows:

```
    src/__tests__/best.test.ts:285:11 - error TS2322: Type 'BestResult<LengthUnits>' is not assignable to type 'LengthUnits'.

    285     const _units: LengthUnits = best; // oops; should be best.unit on the RHS
```

In contrast, I would choose to write the test like this:

```typescript
test('type of unit field matches configured units', () => {
  const convert = configureMeasurements<'length', LengthSystems, LengthUnits>({
    length,
  });
  const best = convert(1200).from('mm').toBest();
  if (best !== null) {
    const _units: LengthUnits = best.unit; // this should type-check
  }
});
```

Just like all the other tests in this file (which I quite like), it exercises the public API, and verifies that the behavior—in this case, the typing behavior—is correct. It uses only the concepts with which the reader is already familiar, and also fits in well with the tests that come right before it: you don't have to switch your brain from "convert-units mode" to "type algebra mode". As to the linter complaint, [`@typescript-eslint/unused-vars`] describes how to configure it to allow variable "names starting with `_` [following] the TypeScript style".

For context, I have been writing typing tests like this for about 6 years. I find it unobtrusive and easily maintainable; it works in small and large codebases, and with people who are very familiar with TypeScript/Flow/etc. as well as with those who are new. (Citations available upon request.)

[`@typescript-eslint/unused-vars`]: https://typescript-eslint.io/rules/no-unused-vars/

If after considering this you find that you would prefer to write the test with `expect-type` and as absolutely little runtime code as possible, I will update this patch accordingly. At the end of the day, I would rather ship this improvement to the package even if I might have written the tests differently myself—and your opinions as maintainer should and do carry more weight than mine! But before I start editing again, I would appreciate hearing your thoughts in response, too. Thanks for bearing with me! <3

Comment (wchargin) [reply src/__tests__/best.test.ts]: Okay: I have switched to the style of test that you requested. Please take a look at your convenience.

> Maybe a different library could produce more user-friendly messages, but I don't know if that would be possible given that the complexity of the error comes from Typescript's understanding of the types. In addition, since both styles rely on Typescript's understanding of the types, it can't be solely an issue with only one style of writing tests.

For what it's worth: this complexity does in fact stem directly from the `expect-type` library. Both rely on TypeScript's understanding of the types, yes, but only one of them *additionally* depends on a 1200-line library with [deeply nested type deconstructors like this][1] (which is what yields the long and unhelpful message compared to tsc's built-in output).

[1]: https://github.com/mmkal/expect-type/blob/89fa7a5a06339d45d468184c891cffb638a50815/src/index.ts#L100-L115

> I want you to know that I didn't enter this discussion to get you to completely change your mind. I did so in hopes to learn something new, explore new ideas, and if given enough reason to, change my position.

Certainly; this is clear and I never got the impression otherwise. No hard feelings!

**Review by Taar — APPROVED** [review]
Thank you for finding this improvement and taking the time to submit a pull request. Greatly appreciated! :)

**Pull request merged**
